### ISSUE TITLE
Know if a length value is 0 independent of its unit?

### ISSUE THREAD

**Issue by blixt** [sass/sass#1846]
In CSS, any length (`px`, `em`, `%`, etc) can be specified as `0` (note that for example angles cannot, so `rotate(0)` is invalid and has to be `rotate(0turn)` or some such).

Right now I'm using this to see if a value is positive or if it's `0`/`0px`/negative:

``` sass
$value = if(0 < $x, $positive, $zero-or-negative);
```

The reason I use `0 < $x` and not `$x != 0` or other variants is because `0 > 0px` and `0px > 0` are both true. The logic around zero in general is seemingly inconsistent or at least confusing, so my proposal is that there is a way to know if a length is zero (or non-zero), something like:

``` sass
$value = if(zero($x), $zero, $non-zero);
```

My use case for this is that some functions (e.g., `random(x)`) require a positive value but I want my function to take both `0` and `0px` (both valid length values) if no randomness is desired. The problem I'm encountering is that there is no reliable way to avoid calling `random` without several checks based on the unit of the length.

If a better way already exists, please let me know. :)

**Comment by cimmanon** [comment]
See https://github.com/sass/sass/issues/1496

**Comment by blixt** [comment]
@cimmanon I read through it and I agree with the general sentiment that you shouldn't compare stuff like `5em == 5` because it doesn't semantically make sense. I think that the concept of zero warrants something that is not a mathematical operation (because of the additional issues that causes around equality), but rather a function such as `zero(0px) == zero(0)`.

The reason I think zero makes sense as a special case is because it is a special case in CSS (`border-width: 0;`, `transform: translate(0, 3px);` etc).

**Comment by davidkpiano** [comment]
@blixt I tend to agree but the W3C only [explicitly allows unitless values for lengths](http://www.w3.org/TR/css3-values/#lengths):

> However, for zero lengths the unit identifier is optional (i.e. can be syntactically represented as the <number> 0).

Given this, most browsers only permit optional unit identifiers for **only length values**, and not other values, such as time values.

``` scss
// This is invalid CSS in most browsers
.test {
  animation-delay: 0;
}

// This is valid
.test {
  animation-delay: 0s;
}

// Therefore...
$zero-time-values-equal: 0s == 0;
// => false (which should be expected)
```

**Comment by blixt** [comment]
@davidkpiano Yep I agree, hence my note regarding degrees which just like times may not be unitless. That said, lengths are one of the most common types of values that you work with in CSS so I still see the benefit of being able to check for zero.

If there is a risk of abuse of this function with non-length values, would it be possible to make it more restrictive in specification so that it may only apply to lengths and unitless values? For example, `zero-length(0)` and `zero-length(0px)` work, but `zero-length(0deg)` and `zero-length(0s)` throw an error? I'm not sure if it's needed and don't really prefer it as it adds complexity, but it seems like a solvable problem (if it is a conceivable problem in the first place).

**Comment by chriseppstein** [comment]
Sass doesn't know the difference between seconds and pixels. Sass just knows that these units are the same or different and it knows which units can be converted to other units. We do this for a reason: future compatibility. Each time Sass has a behavior that must change when CSS specs change, we create a situation where old versions of Sass cannot behave as expected (until upgrading, which may or may not be possible for your project at that point in time).

TL;DR I think this is a good use case for a custom function.

**Comment by blixt** [comment]
I understand your concern, but this is about a behavior that hasn't changed for almost 20 years. I created the issue to solely discuss the fact that units on the 0 length are optional (so sorry about the sidetracking regarding time units, etc.), and this has been the case in the CSS specification, without change, [since 1996](http://www.w3.org/TR/REC-CSS1-961217#length-units).

I'm using a custom function for now, but I see no reason why there shouldn't be a way to compare lengths with a unitless 0 since it's such an integral part of CSS since version 1.
